File: bindings/python/src/tapirscan/formats.py
"""Explicit opt-in barcode formats, matching the native ABI."""

from typing import Literal, TypeAlias, cast

Format: TypeAlias = Literal[
    "EAN13",
    "UPCA",
    "EAN8",
    "UPCE",
    "Code128",
    "Code39",
    "ITF",
    "Codabar",
    "Code93",
    "QRCode",
    "DataMatrix",
    "PDF417",
    "Aztec",
    "DataBar",
    "DataBarExpanded",
    "MaxiCode",
]
FORMAT_BITS: dict[str, int] = {
    "EAN13": 1,
    "UPCA": 2,
    "EAN8": 4,
    "UPCE": 8,
    "Code128": 16,
    "Code39": 32,
    "ITF": 64,
    "Codabar": 128,
    "Code93": 256,
    "QRCode": 512,
    "DataMatrix": 1024,
    "PDF417": 2048,
    "Aztec": 4096,
    "DataBar": 8192,
    "DataBarExpanded": 16384,
    "MaxiCode": 131072,
}


FormatSelection: TypeAlias = (
    Format
    | tuple[Format, ...]
    | list[Format]
    | Literal["retail", "common1D", "common", "1D", "2D", "all"]
)
retail_formats: tuple[Format, ...] = ("EAN13", "UPCA", "EAN8", "UPCE")
common_linear_formats: tuple[Format, ...] = (
    *retail_formats,
    "Code128",
    "Code39",
    "ITF",
)
common_formats: tuple[Format, ...] = (*common_linear_formats, "QRCode", "DataMatrix")
linear_formats: tuple[Format, ...] = (
    *common_linear_formats,
    "Codabar",
    "Code93",
    "DataBar",
    "DataBarExpanded",
)
matrix_formats: tuple[Format, ...] = (
    "QRCode",
    "DataMatrix",
    "PDF417",
    "Aztec",
    "MaxiCode",
)


def resolve_formats(formats: FormatSelection | None) -> tuple[Format, ...]:
    """Expand public presets and validate a nonempty explicit format selection."""
    if formats is None:
        return ("EAN13",)
    if isinstance(formats, str):
        if formats in FORMAT_BITS:
            return (cast("Format", formats),)
        presets = {
            "retail": retail_formats,
            "common1D": common_linear_formats,
            "common": common_formats,
            "1D": linear_formats,
            "2D": matrix_formats,
            "all": (*linear_formats, *matrix_formats),
        }
        if formats not in presets:
            msg = f"Unknown format preset: {formats}"
            raise ValueError(msg)
        return presets[formats]
    if not isinstance(formats, (tuple, list)) or not formats:
        msg = "Choose at least one barcode format"
        raise ValueError(msg)
    for name in formats:
        if not isinstance(name, str) or name not in FORMAT_BITS:
            msg = f"Unsupported barcode format: {name}"
            raise ValueError(msg)
    return tuple(dict.fromkeys(formats))


def format_mask(formats: FormatSelection | None) -> int:
    """Validate formats before native code is called."""
    mask = 0
    for name in resolve_formats(formats):
        mask |= FORMAT_BITS[name]
    return mask

File: bindings/javascript/src/multiformat/formats.ts
/** Explicit opt-in formats for the independent Medium extension. */
import {
  commonFormats,
  commonLinearFormats,
  formatBits,
  linearFormats,
  matrixFormats,
  retailFormats,
  type Format,
} from "./format-registry.js";

export {
  commonFormats,
  commonLinearFormats,
  formatBits,
  linearFormats,
  matrixFormats,
  retailFormats,
  type Format,
};
export type FormatSelection =
  Format | readonly Format[] | "retail" | "common1D" | "common" | "1D" | "2D" | "all";
export function resolveFormats(input?: readonly string[] | string): Format[] {
  if (input === "retail") return [...retailFormats];
  if (input === "common1D") return [...commonLinearFormats];
  if (input === "common") return [...commonFormats];
  if (input === "1D") return [...linearFormats];
  if (input === "2D") return [...matrixFormats];
  if (input === "all") return [...linearFormats, ...matrixFormats];
  if (input === undefined) return [...retailFormats];
  if (typeof input === "string" && Object.hasOwn(formatBits, input)) return [input as Format];
  if (!Array.isArray(input) || input.length === 0)
    throw Error("Choose at least one barcode format.");
  for (const value of input as readonly unknown[]) {
    if (typeof value !== "string") throw Error("Barcode format must be a string.");
    if (!Object.hasOwn(formatBits, value)) throw Error(`Unsupported barcode format: ${value}`);
  }
  return [...new Set(input)] as Format[];
}
export function maskFor(formats: readonly Format[]): number {
  return formats.reduce((mask, f) => mask | formatBits[f], 0);
}
